### PR TITLE
Article settings property when item is getting out of stock.

### PR DESCRIPTION
**Pull request by andreesg**
Based on the Pull request: https://github.com/bluedynamics/bda.plone.orders/pull/31

Use case:
* Webshop administrator doesn't notice that items are getting out of stock.

What was implemented:
* New property for Article settings interface: "default_item_minimum_stock"

### CODE REVIEW

File: src/bda/plone/shop/at.py [hunk 1]
```diff
@@ -130,6 +130,11 @@ def default_item_cart_count_limit(context):
 def default_item_quantity_unit(context):
     return lambda: get_shop_article_settings().default_item_quantity_unit
 
+@implementer(IFieldDefaultProvider)
+@adapter(IBuyable)
+def default_item_minimum_stock(context):
+    return lambda: get_shop_article_settings().default_item_minimum_stock
+
```
Comment (rnixx) [inline src/bda/plone/shop/at.py]: This is only necessary once you provide some kind of MinimumItemStock AT schema extender and adapter which can override the default value from article settings on dedicated context. You can either remove the lines or complete the implementation ;). Same applies to https://github.com/bluedynamics/bda.plone.shop/pull/63/files#diff-244eaa6a40836e7820844d9d27a6420fR67 just related to DX a behavior

**Comment by rnixx** [comment]
Why have you changed all the file permissions? Can you revert this please?

The settings interface is good.

The default value providers in at.py and dx.py are useless without the respective implementations.

**Comment by andreesg** [comment]
Thanks for your feedback. Indeed that will be nice to implement the article settings on a dedicated context. However I'm not sure if this is desirable to be modified in context. 
I will commit the changes and revert the file permissions.

**Comment by rnixx** [comment]
I think it would be useful if limit can be overwritten on dedicated context. There might be the case that a shop admin wants to gets informed earlier for items which are sold more frequently or which take longer to purchase

**Comment by andreesg** [comment]
Yes, very good point, that would be really useful. Is this something that should go on the IStockBehavior ? Therefore the ICartItemStock should also be updated - and the bring back the default value providers in at.py and dx.py. Am I missing something?

**Comment by rnixx** [comment]
``bda.plone.cart.interfaces.ICartItemStockICartItemStock`` is the relevant interface, there you can add the attribute. Then it needs to be implemented in ``bda.plone.shop.at.StockExtender``, ``bda.plone.shop.at.ATCartItemStock``, ``bda.plone.shop.dx.IStockBehavior`` and ``bda.plone.shop.dx.DXCartItemStock``. Also the respective default value providers should be re-added. The send out logic should then acquire the value from ICartItemStock adapter.

**Comment by andreesg** [comment]
The interfaces were updated in order to support the minimum stock in context.

bda.plone.orders and bda.plone.cart were updated accordingly:
* https://github.com/bluedynamics/bda.plone.cart/pull/21
* https://github.com/bluedynamics/bda.plone.orders/pull/31/commits/f245d7dde99e849d53d3581ba6c1fa04ab8c717b

Please let me know what you think and if it is ready to merge.

**Comment by rnixx** [comment]
The implementation looks complete so far in all 3 pull requests.

It would be awesome if you can improve the naming a little bit. We are neither "out of stock" nor do we have a "minimum stock".

In my understanding we have "low stock reached" and a "stock warning threshold" or so..

anyone with a better naming suggestion?

**Comment by andreesg** [comment]
Yes. I will improve the naming. 

"stock warning threshold" looks good.

**Comment by thet** [comment]
+1 for ``stock_warning_threshold``.

**Comment by andreesg** [comment]
The names were improved in all 3 Pull requests.

The name of the field in the StockBehavior interface is now: `stock_warning_threshold`
The Event that is notified when threshold is reached is now: `StockThresholdReached` that implements the interface `IStockThresholdReached`

Thanks for your help! are we ready ?

File: src/bda/plone/shop/at.py [hunk 2]
```diff
@@ -323,6 +336,15 @@ def _set_overbook(self, value):
 
     overbook = property(_get_overbook, _set_overbook)
 
+    def _get_stock_warning_threshold(self):
+        return field_value(self.context, 'stock_warning_threshold')
```
Comment (rnixx) [inline src/bda/plone/shop/at.py]: Wrong fieldname, must be 'item_stock_warning_threshold'

File: src/bda/plone/shop/at.py [hunk 3]
```diff
@@ -323,6 +336,15 @@ def _set_overbook(self, value):
...
+
+    def _set_stock_warning_threshold(self, value):
+        set_field_value(self.context, 'stock_warning_threshold', value)
```
Comment (rnixx) [inline src/bda/plone/shop/at.py]: Wrong fieldname, must be 'item_stock_warning_threshold'

**Comment by rnixx** [comment]
Please revert file mode change of i18n.sh. We're almost there ;)

**Comment by thet** [comment]
Upgrade step is missing!
I just got the error:
``KeyError: 'Interface `bda.plone.shop.interfaces.IShopArticleSettings` defines a field `default_item_stock_warning_threshold`, for which there is no record.'``

**Comment by thet** [comment]
Also, changelog was missing. Both added here: https://github.com/bluedynamics/bda.plone.shop/commit/63eed6b7647b949413dedb0f9318a403d1d54a96

**Pull request merged**